### PR TITLE
New auth system for rcon.

### PR DESCRIPTION
**Pull request by Learath2**
Turns out @heinrich5991 is hard to please /s

This time with salted&hashed passwords and ugly teeworlds braces :)

### CODE REVIEW

**Comment by heinrich5991** [comment]
Thanks for the update!

It seems the server executable crashes on startup:
```
$ ./DDNet-Server_d 
'./DDNet-Server_d' terminated by signal SIGSEGV (Address boundary error)
$ ./DDNet-Server_d 
[17-03-03 17:13:29][engine]: running on unix-linux-amd64
[17-03-03 17:13:29][engine]: arch is little endian
[17-03-03 17:13:29][storage]: added path '$USERDIR' ('/home/hein/.teeworlds')
[17-03-03 17:13:29][storage]: added path '$DATADIR' ('data')
[17-03-03 17:13:29][storage]: added path '$CURRENTDIR' ('/home/hein/git/teeworlds')
[17-03-03 17:13:29][host lookup]: host='localhost' port=0 1
[17-03-03 17:13:29][host lookup]: host='localhost' port=0 2
[17-03-03 17:13:29][console]: executing 'autoexec_server.cfg'
[17-03-03 17:13:29][assert]: src/engine/shared/network_server.cpp(678): errornous client id
'./DDNet-Server_d' terminated by signal SIGSEGV (Address boundary error)
```

File: src/engine/server/server.cpp [hunk 1]
```diff
@@ -2230,7 +2418,18 @@ void CServer::ConchainRconPasswordChange(IConsole::IResult *pResult, void *pUser
...
+		CAuthManager *pManager = &pServer->m_AuthManager;
+
+		int KeySlot = pManager->DefaultKey(AUTHED_ADMIN);
+		if(KeySlot == -1)
+		{
+			pManager->AddAdminKey(pResult->GetString(0));//Shouldn't happen
+		}
+		else
+		{
+			pManager->UpdateKey(KeySlot, pResult->GetString(0), AUTHED_ADMIN);
```
Comment (heinrich5991) [inline src/engine/server/server.cpp]: This should remove the key if the new rcon password is empty.

File: src/game/client/components/console.cpp [hunk 2]
```diff
@@ -470,7 +487,17 @@ void CGameConsole::OnRender()
...
 					pPrompt = "rcon> ";
 				else
-					pPrompt = "ENTER PASSWORD> ";
+				{
+					if(IsDDNet(&pServerInfo))
+					{
+						if(!pConsole->m_aUser[0])
+							pPrompt = "Enter Username> ";
+						else
+							pPrompt = "Enter Password> ";
```
Comment (Learath2) [inline src/game/client/components/console.cpp]: I had to compromise there b/c there is no way to let the client know there is no extended users. I could modify the routine for the return key to allow empty usernames but that looks ugly.

default_admin is the ident for sv_rcon_password

Comment (east) [inline src/game/client/components/console.cpp]: The protocol extension would become handy here, the server can tell the client whether it is supposed to use name&password or password only.

Comment (Learath2) [inline src/game/client/components/console.cpp]: I already have a branch where I use this but wanted to get this out asap.

File: src/engine/server/server.cpp [hunk 3]
```diff
@@ -2230,7 +2395,21 @@ void CServer::ConchainRconPasswordChange(IConsole::IResult *pResult, void *pUser
 	if(pResult->NumArguments() == 1)
 	{
 		CServer *pServer = (CServer *)pUserData;
-		pServer->LogoutByAuthLevel(AUTHED_ADMIN);
+		CAuthManager *pManager = &pServer->m_AuthManager;
+
+		int KeySlot = pManager->DefaultKey(AUTHED_ADMIN);
+		if(KeySlot == -1)
+		{
+			pManager->AddAdminKey(pResult->GetString(0));//Shouldn't happen except for the first launch
```
Comment (heinrich5991) [inline src/engine/server/server.cpp]: And also happens if the key is removed. Just remove the comment? :P

File: src/engine/server/server.cpp [hunk 4]
```diff
@@ -1944,6 +1950,167 @@ void CServer::ConDnsblStatus(IConsole::IResult *pResult, void *pUser)
...
+	{
+		pThis->Console()->Print(IConsole::OUTPUT_LEVEL_STANDARD, "auth", "level can be one of {\"admin\", \"mod(erator)\", \"helper\"}");
+		return;
+	}
+
+	unsigned char aHash[MD5_BYTES];
+	unsigned char aSalt[SALT_BYTES];
+
+	str_hex_decode(aHash, sizeof aHash, pPw);
+	str_hex_decode(aSalt, sizeof aSalt, pSalt);
```
Comment (heinrich5991) [inline src/engine/server/server.cpp]: Missing error handling if `pPw` or `pSalt` don't contain valid hex.

File: src/engine/server/server.cpp [hunk 5]
```diff
@@ -1944,6 +1950,167 @@ void CServer::ConDnsblStatus(IConsole::IResult *pResult, void *pUser)
...
```
Comment (heinrich5991) [inline src/engine/server/server.cpp]: Also missing error handling.

File: src/engine/server/server.cpp [hunk 6]
```diff
@@ -770,6 +759,7 @@ int CServer::ClientRejoinCallback(int ClientID, void *pUser)
 	CServer *pThis = (CServer *)pUser;
 
 	pThis->m_aClients[ClientID].m_Authed = AUTHED_NO;
+	pThis->m_aClients[ClientID].m_AuthKey = 0;
```
Comment (heinrich5991) [inline src/engine/server/server.cpp]: `m_AuthKey` being 0 already means that the client is authenticated with key 0, I think this should be -1 (same below).

File: src/engine/server/server.cpp [hunk 7]
```diff
@@ -1906,8 +1908,10 @@ void CServer::ConStatus(IConsole::IResult *pResult, void *pUser)
 				const char *pAuthStr = pThis->m_aClients[i].m_Authed == CServer::AUTHED_ADMIN ? "(Admin)" :
 										pThis->m_aClients[i].m_Authed == CServer::AUTHED_MOD ? "(Mod)" :
 										pThis->m_aClients[i].m_Authed == CServer::AUTHED_HELPER ? "(Helper)" : "";
+				char aAuthStr[128];
+				str_format(aAuthStr, sizeof aAuthStr, "key=%s %s", pThis->m_AuthManager.KeyIdent(pThis->m_aClients[i].m_AuthKey), pAuthStr);
```
Comment (heinrich5991) [inline src/engine/server/server.cpp]: This doesn't work if the client isn't logged in.

File: src/engine/server/server.cpp [hunk 8]
```diff
@@ -1934,8 +1938,10 @@ void CServer::ConDnsblStatus(IConsole::IResult *pResult, void *pUser)
...
```
Comment (heinrich5991) [inline src/engine/server/server.cpp]: Same as above.

File: src/game/client/components/console.cpp (hunk 2 above)
Comment (heinrich5991) [inline src/game/client/components/console.cpp]: OK, then we only need a "temporary" solution. Would just pressing enter on the username work with you?

File: src/base/system.c [hunk 9]
```diff
@@ -2146,6 +2146,57 @@ void str_hex(char *dst, int dst_size, const void *data, int data_size)
...
+    case 'E': return 14;
+    case 'f':
+    case 'F': return 15;
+    default: return -1;
+    }
+}
+
+static unsigned char byteval(const char *byte)
+{
+	return hexval(byte[0]) * 16 + hexval(byte[1]);
```
Comment (Learath2) [inline src/base/system.c]: Fixed, but its called in a controlled environment so not sure if its even needed.

**Comment by Learath2** [comment]
Huh I'm not sure how that even worked at all.

**Comment by heinrich5991** [comment]
Thanks a lot, @Learath2!

**Comment by timakro** [comment]
Is just setting `sv_rcon_password` still compatible with vanilla clients?

**Comment by heinrich5991** [comment]
Unless there's a bug, yes, it's still compatible with vanilla clients.

**Pull request merged**